### DISCUSSION TITLE
Cask blackhole-2ch uses a command that is not in the latest cask code

### DISCUSSION THREAD

**Discussion by T3ll1nG** [Homebrew/discussions#5291] [Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.18-29-g5517894
ORIGIN: https://github.com/Homebrew/brew
HEAD: 5517894523620d2000f96013aa770b51dd73effc
Last commit: 15 hours ago
Core tap HEAD: 7dc6d5c20d0c9b15b9092446687a9d7ce6fac50e
Core tap last commit: 13 minutes ago
Core tap JSON: 16 Apr 15:03 UTC
Core cask tap HEAD: 7af16bf6544146cbc3d90841076dcc4594004144
Core cask tap last commit: 2 hours ago
Core cask tap JSON: 16 Apr 15:03 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.44.0 => /opt/homebrew/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.4.1-arm64
CLT: 15.3.0.0.1.1708646388
Xcode: 15.3
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

When I run `brew upgrade` on blackhole-2ch the Terminal has the following output:

> ==> Uninstalling packages with sudo; the password may be necessary:
> Could not kickstart service "com.apple.audio.coreaudiod": 1: Operation not permitted
> ==> Purging files for version 0.6.0 of Cask blackhole-2ch
> Error: blackhole-2ch: Failure while executing; `/usr/bin/sudo -E -- /bin/launchctl kickstart -kp system/com.apple.audio.coreaudiod` exited with 1. Here's the output:
> Could not kickstart service "com.apple.audio.coreaudiod": 1: Operation not permitted

This is a old problem since `kickstart` is no longer [permitted](https://developer.apple.com/documentation/macos-release-notes/macos-14_4-release-notes#Core-Audio).

But this is fixed in the latest [cask](https://github.com/Homebrew/homebrew-cask/blob/7af16bf6544146cbc3d90841076dcc4594004144/Casks/b/blackhole-2ch.rb) and in fact it was [fixed](https://github.com/Homebrew/homebrew-cask/commit/d6de09ea49580c5e73fcf3db6c35f18249affb28) a month ago by @krehel.

Obviously something is wrong but the update wants to install 0.6 which is the correct version, so I do not get how it could call code that is no longer present.

### Relevant casks

blackhole-2ch

**Comment by SMillerDev** [comment]
Afaik uninstall is based on previously installed version of the cask.

**Comment by T3ll1nG** [comment, marked as answer]
@SMillerDev  was right. The `.rb` is stored in `/opt/homebrew/Caskroom/Caskname/.metadata/` and in fact uses the code for the version you have installed.

Here is a short script you can use to replace it if you are on `version "0.5.0"`. Check for potential issues and if none are found you can delete the copy of the original on your Desktop.

```
#!/bin/zsh

# Make a copy of the original file on the desktop
cp $(brew --prefix)/Caskroom/blackhole-2ch/.metadata/0.5.0/20230214075027.489/Casks/blackhole-2ch.rb ~/Desktop/blackhole-2ch_original.rb

# Replace lines 19 to 28 in the file
sed -i '' '19,28d' $(brew --prefix)/Caskroom/blackhole-2ch/.metadata/0.5.0/20230214075027.489/Casks/blackhole-2ch.rb

# Insert the new code block after line 18
sed -i '' '18 a\
    uninstall_postflight do\
        system_command "/usr/bin/killall",\
                    args:         ["coreaudiod"],\
                    sudo:         true,\
                    must_succeed: true\
    end
' $(brew --prefix)/Caskroom/blackhole-2ch/.metadata/0.5.0/20230214075027.489/Casks/blackhole-2ch.rb

# Open the file for verification
open $(brew --prefix)/Caskroom/blackhole-2ch/.metadata/0.5.0/20230214075027.489/Casks/blackhole-2ch.rb
```

**Reply by krehel** [reply]
Note to others: this would also apply to `blackhole-16ch` and `blackhole-64ch` as the same modifications are present in those Casks.

**Reply by Iksas** [reply]
Note that on Intel Macs, the `Caskroom` folder is located at `/usr/local/Caskroom` instead of `/opt/homebrew/Caskroom`.

**Reply by krehel** [reply]
Thanks @Iksas - you're right - ultimately should use `$(brew --prefix)/Caskroom` to obtain this as that will handle the Intel or AS appropriate pathing.

**Reply by T3ll1nG** [reply]
Thanks @krehel for mentioning `$(brew --prefix)/Caskroom` I changed the script according to that post. Also thanks for @Iksas for pointing to the fact that there are two possible locations for brew.

**Reply by marktmiller** [reply]
I'm getting the same thing.

I've tried uninstalling Ver. 0.6.0, and I can't seem to clean it out. I've tried brew uninstall --force blackhole-2ch, and when I try to install Ver. 0.5.0, brew says "newest version already installed," ref. to Ver. 0.6.0.

The answer posted here talks about what to do if you have Ver. 0.5.0. What should people who've gotten the botched 0.6.0 upgrade do?

Thanks.
